### Student Information
Name: Hsin-Ju Li(李信儒)

Student ID: 110033630

GitHub ID: popo6699325

Kaggle name: Hsin-Ju,li

Kaggle private scoreboard snapshot:

[Snapshot](img/pic0.png)

---

### Instructions

1. First: __This part is worth 30% of your grade.__ Do the **take home** exercises in the [DM2021-Lab2-master Repo](https://github.com/fhcalderon87/DM2021-Lab2-master). You may need to copy some cells from the Lab notebook to this notebook. 


2. Second: __This part is worth 30% of your grade.__ Participate in the in-class [Kaggle Competition](https://www.kaggle.com/c/dm2021-lab2-hw2/) regarding Emotion Recognition on Twitter. The scoring will be given according to your place in the Private Leaderboard ranking: 
    - **Bottom 40%**: Get 20% of the 30% available for this section.

    - **Top 41% - 100%**: Get (60-x)/6 + 20 points, where x is your ranking in the leaderboard (ie. If you rank 3rd your score will be (60-3)/6 + 20 = 29.5% out of 30%)   
    Submit your last submission __BEFORE the deadline (Dec. 24th 11:59 pm, Friday)__. Make sure to take a screenshot of your position at the end of the competition and store it as '''pic0.png''' under the **img** folder of this repository and rerun the cell **Student Information**.
    

3. Third: __This part is worth 30% of your grade.__ A report of your work developping the model for the competition (You can use code and comment it). This report should include what your preprocessing steps, the feature engineering steps and an explanation of your model. You can also mention different things you tried and insights you gained. 


4. Fourth: __This part is worth 10% of your grade.__ It's hard for us to follow if your code is messy :'(, so please **tidy up your notebook** and **add minimal comments where needed**.


Upload your files to your repository then submit the link to it on the corresponding e-learn assignment.

Make sure to commit and save your changes to your repository __BEFORE the deadline (Dec. 29th 11:59 pm, Wednesday)__. 

In [2]:
### Begin Assignment Here

### 2. Second : Kaggle Competition

#### Data Classification

In [1]:
import pandas as pd

tweets_DM_df = pd.read_json("tweets_DM.json", lines=True)
emotion_df = pd.read_csv("emotion.csv")
data_identification_df = pd.read_csv("data_identification.csv")

In [2]:
tweets_DM_df

_score          _index  \
0           391  hashtag_tweets   
1           433  hashtag_tweets   
2           232  hashtag_tweets   
3           376  hashtag_tweets   
4           989  hashtag_tweets   
...         ...             ...   
1867530     827  hashtag_tweets   
1867531     368  hashtag_tweets   
1867532     498  hashtag_tweets   
1867533     840  hashtag_tweets   
1867534     360  hashtag_tweets   

                                                   _source  \
0        {'tweet': {'hashtags': ['Snapchat'], 'tweet_id...   
1        {'tweet': {'hashtags': ['freepress', 'TrumpLeg...   
2        {'tweet': {'hashtags': ['bibleverse'], 'tweet_...   
3        {'tweet': {'hashtags': [], 'tweet_id': '0x1cd5...   
4        {'tweet': {'hashtags': [], 'tweet_id': '0x2de2...   
...                                                    ...   
1867530  {'tweet': {'hashtags': ['mixedfeeling', 'butim...   
1867531  {'tweet': {'hashtags': [], 'tweet_id': '0x29d0...   
1867532  {'tweet': {'hashtags': [], 'tweet_id': '0x2a6a...   
1867533  {'tweet': {'hashtags': [], 'tweet_id': '0x24fa...   
1867534  {'tweet': {'hashtags': ['Sundayvibes'], 'tweet...   

                  _crawldate   _type  
0        2015-05-23 11:42:47  tweets  
1        2016-01-28 04:52:09  tweets  
2        2017-12-25 04:39:20  tweets  
3        2016-01-24 23:53:05  tweets  
4        2016-01-08 17:18:59  tweets  
...                      ...     ...  
1867530  2015-05-12 12:51:52  tweets  
1867531  2017-10-02 17:54:04  tweets  
1867532  2016-10-10 11:04:32  tweets  
1867533  2016-09-02 14:25:06  tweets  
1867534  2016-11-16 01:40:07  tweets  

[1867535 rows x 5 columns]

In [3]:
emotion_df

tweet_id       emotion
0        0x3140b1       sadness
1        0x368b73       disgust
2        0x296183  anticipation
3        0x2bd6e1           joy
4        0x2ee1dd  anticipation
...           ...           ...
1455558  0x38dba0           joy
1455559  0x300ea2           joy
1455560  0x360b99          fear
1455561  0x22eecf           joy
1455562  0x2fb282  anticipation

[1455563 rows x 2 columns]

In [4]:
data_identification_df

tweet_id identification
0        0x28cc61           test
1        0x29e452          train
2        0x2b3819          train
3        0x2db41f           test
4        0x2a2acc          train
...           ...            ...
1867530  0x227e25          train
1867531  0x293813          train
1867532  0x1e1a7e          train
1867533  0x2156a5          train
1867534  0x2bb9d2          train

[1867535 rows x 2 columns]

In [5]:
hashtags = []
id = []
text = []
for source in tweets_DM_df['_source']:
    hashtags.append(source['tweet']['hashtags'])
    id.append(source['tweet']['tweet_id'])
    text.append(source['tweet']['text'])
tweets_DM_df['hashtags'] = hashtags
tweets_DM_df['tweet_id'] = id
tweets_DM_df['text'] = text


In [6]:
tweets_DM_df.drop(columns='_index', inplace=True)
tweets_DM_df.drop(columns='_source', inplace=True)
tweets_DM_df.drop(columns='_crawldate', inplace=True)
tweets_DM_df.drop(columns='_type', inplace=True)

In [7]:
tweets_DM_df

_score                         hashtags  tweet_id  \
0           391                       [Snapchat]  0x376b20   
1           433    [freepress, TrumpLegacy, CNN]  0x2d5350   
2           232                     [bibleverse]  0x28b412   
3           376                               []  0x1cd5b0   
4           989                               []  0x2de201   
...         ...                              ...       ...   
1867530     827  [mixedfeeling, butimTHATperson]  0x316b80   
1867531     368                               []  0x29d0cb   
1867532     498                               []  0x2a6a4f   
1867533     840                               []  0x24faed   
1867534     360                    [Sundayvibes]  0x34be8c   

                                                      text  
0        People who post "add me on #Snapchat" must be ...  
1        @brianklaas As we see, Trump is dangerous to #...  
2        Confident of your obedience, I write to you, k...  
3                      Now ISSA is stalking Tasha 😂😂😂 <LH>  
4        "Trust is not the same as faith. A friend is s...  
...                                                    ...  
1867530  When you buy the last 2 tickets remaining for ...  
1867531  I swear all this hard work gone pay off one da...  
1867532  @Parcel2Go no card left when I wasn't in so I ...  
1867533  Ah, corporate life, where you can date <LH> us...  
1867534             Blessed to be living #Sundayvibes <LH>  

[1867535 rows x 4 columns]

In [8]:
test_df=data_identification_df[data_identification_df['identification'] == 'test']
train_df=data_identification_df[data_identification_df['identification'] == 'train']
train_data_df = tweets_DM_df[tweets_DM_df.tweet_id.isin(train_df.tweet_id)].reset_index(drop=True)
test_data_df = tweets_DM_df[tweets_DM_df.tweet_id.isin(test_df.tweet_id)].reset_index(drop=True)

In [9]:
train_data_df = train_data_df.sort_values(by=['tweet_id']).reset_index(drop=True)
emotion_df = emotion_df.sort_values(by=['tweet_id']).reset_index(drop=True)

In [10]:
train_data_df['emotion'] = emotion_df.emotion

In [11]:
train_data_df

_score                                           hashtags  tweet_id  \
0           242                                      [BlackMirror]  0x1c7f10   
1           915  [twitch, Destinybeta, Destiny, Destiny2, Desti...  0x1c7f11   
2           939                                                 []  0x1c7f14   
3           181  [Confession, NationalCandyCornDay, CouldEatThe...  0x1c7f15   
4           970                                                 []  0x1c7f16   
...         ...                                                ...       ...   
1455558     922                                                 []  0x38fe18   
1455559      77                     [WesHoolahan, WALvIRL, COYBIG]  0x38fe19   
1455560      25                                        [not, maga]  0x38fe1a   
1455561     639                                                 []  0x38fe1c   
1455562     630                     [redheadproblems, ouch, burnt]  0x38fe1d   

                                                      text       emotion  
0         o m g Shut Up And Dance though #BlackMirror <LH>           joy  
1        On #twitch <LH> on the #Destinybeta #Destiny #...  anticipation  
2        A nice sunny wak this morning not many <LH> ar...           joy  
3        I'm one of those people who love candy corn......           joy  
4        @metmuseum What are these? They look like some...       disgust  
...                                                    ...           ...  
1455558  @LJPBR @FifthHarmony Um  My vote For @FifthHar...       sadness  
1455559     Where is #WesHoolahan?!  #WALvIRL #COYBIG <LH>  anticipation  
1455560  @mattmfm Fake news! <LH> propagated by Tumpkin...      surprise  
1455561                    ..today was brutal  ..#Hungover       disgust  
1455562  Love it when I sun burn my forehead!! NOT!! 😫😱...       sadness  

[1455563 rows x 5 columns]

In [12]:
test_data_df

_score                           hashtags  tweet_id  \
0          232                       [bibleverse]  0x28b412   
1          989                                 []  0x2de201   
2           66  [materialism, money, possessions]  0x218443   
3          104               [GodsPlan, GodsWork]  0x2939d5   
4          310                                 []  0x26289a   
...        ...                                ...       ...   
411967     602                                 []  0x2913b4   
411968     598                                 []  0x2a980e   
411969     827    [mixedfeeling, butimTHATperson]  0x316b80   
411970     368                                 []  0x29d0cb   
411971     498                                 []  0x2a6a4f   

                                                     text  
0       Confident of your obedience, I write to you, k...  
1       "Trust is not the same as faith. A friend is s...  
2       When do you have enough ? When are you satisfi...  
3       God woke you up, now chase the day #GodsPlan #...  
4       In these tough times, who do YOU turn to as yo...  
...                                                   ...  
411967  "For this is the message that ye heard from th...  
411968  "There is a lad here, which hath five barley l...  
411969  When you buy the last 2 tickets remaining for ...  
411970  I swear all this hard work gone pay off one da...  
411971  @Parcel2Go no card left when I wasn't in so I ...  

[411972 rows x 4 columns]

#### Data process

In [13]:
train_data_df.groupby(['emotion']).count()['text']

emotion
anger            39867
anticipation    248935
disgust         139101
fear             63999
joy             516017
sadness         193437
surprise         48729
trust           205478
Name: text, dtype: int64

##### Undersampling (All class size are same as anger)

In [52]:
import numpy as np
arr_train_anger = np.array(train_data_df[train_data_df.emotion == 'anger'])
arr_train_anticipation = np.array(train_data_df[train_data_df.emotion == 'anticipation'])
arr_train_disgust = np.array(train_data_df[train_data_df.emotion == 'disgust'])
arr_train_fear = np.array(train_data_df[train_data_df.emotion == 'fear'])
arr_train_joy = np.array(train_data_df[train_data_df.emotion == 'joy'])
arr_train_sadness = np.array(train_data_df[train_data_df.emotion == 'sadness'])
arr_train_surprise = np.array(train_data_df[train_data_df.emotion == 'surprise'])
arr_train_trust = np.array(train_data_df[train_data_df.emotion == 'trust'])

In [53]:
np.random.shuffle(arr_train_anger)
np.random.shuffle(arr_train_anticipation)
np.random.shuffle(arr_train_disgust)
np.random.shuffle(arr_train_fear)
np.random.shuffle(arr_train_joy)
np.random.shuffle(arr_train_sadness)
np.random.shuffle(arr_train_surprise)
np.random.shuffle(arr_train_trust)

In [54]:
size = len(arr_train_anger)
new_arr_train_anger = arr_train_anger[:size,:]
new_arr_train_anticipation = arr_train_anticipation[:size,:]
new_arr_train_disgust = arr_train_disgust[:size,:]
new_arr_train_fear = arr_train_fear[:size,:]
new_arr_train_joy = arr_train_joy[:size,:]
new_arr_train_sadness = arr_train_sadness[:size,:]
new_arr_train_surprise = arr_train_surprise[:size,:]
new_arr_train_trust = arr_train_trust[:size,:]

In [55]:
columns = train_data_df.columns

In [56]:
train_data_df = np.concatenate((new_arr_train_anger, new_arr_train_anticipation, new_arr_train_disgust, new_arr_train_fear, new_arr_train_joy,
                                new_arr_train_sadness, new_arr_train_surprise, new_arr_train_trust), axis=0)

In [57]:
np.random.shuffle(train_data_df)
train_data_df = pd.DataFrame(train_data_df, columns=columns)

##### Text tokenize

In [14]:
from sklearn.feature_extraction.text import CountVectorizer
import nltk
BOW_vectorizer = CountVectorizer(tokenizer=nltk.word_tokenize)
BOW_vectorizer.fit(train_data_df.text)

train_data_BOW_features = BOW_vectorizer.transform(train_data_df.text)
test_data_BOW_features = BOW_vectorizer.transform(test_data_df.text)


C:\Users\popo6\Anaconda3\envs\datamining\lib\site-packages\sklearn\feature_extraction\text.py:516: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [15]:
print(train_data_BOW_features.shape)
print(test_data_BOW_features.shape)

(1455563, 994692)
(411972, 994692)


In [16]:
feature_name = BOW_vectorizer.get_feature_names_out()
feature_name[:200]

array(['!', '#', '$', '%', '&', "'", "''", "''alhamdulillah", "''and",
       "''be", "''behold", "''believe", "''brands", "''care", "''darlin",
       "''do", "''dream", "''encounters", "''epic", "''family", "''folks",
       "''for", "''funnel", "''have", "''if", "''in", "''know", "''learn",
       "''live", "''looking", "''man", "''me", "''new", "''not", "''now",
       "''others", "''provide", "''say", "''so", "''sometimes",
       "''thanks", "''that", "''the", "''there", "''they", "''this",
       "''tis", "''to", "''told", "''twas", "''very", "''when", "''where",
       "''whose", "''you", "'-", "'-d", "'-mahek.dave", "'-not", "'/",
       "'00", "'000s", "'007", "'02", "'03", "'04", "'04/", "'05", "'06",
       "'06—tv", "'07", "'08", "'09", "'0wner_assad", "'10", "'101",
       "'10k", "'11", "'12", "'13", "'13/9/2017", "'14", "'14th", "'15",
       "'16", "'17", "'17-18", "'17/'18", "'18", "'19", "'1984", "'19s",
       "'1st", "'20", "'20/20", "'2017", "'21", "'22", "'24", "

#### Decision Trees

In [17]:
from sklearn.tree import DecisionTreeClassifier

X_train = train_data_BOW_features
Y_train = train_data_df.emotion

X_test = test_data_BOW_features

clf_decisiontree = DecisionTreeClassifier(max_depth=10000)
clf_decisiontree = clf_decisiontree.fit(X_train, Y_train)

Y_train_pred = clf_decisiontree.predict(X_train)
Y_test_pred = clf_decisiontree.predict(X_test)

In [18]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

print(accuracy_score(y_true=Y_train, y_pred=Y_train_pred))
print(classification_report(y_true=Y_train, y_pred=Y_train_pred))

0.9964803996803986
              precision    recall  f1-score   support

       anger       1.00      1.00      1.00     39867
anticipation       1.00      1.00      1.00    248935
     disgust       1.00      1.00      1.00    139101
        fear       1.00      0.99      0.99     63999
         joy       0.99      1.00      1.00    516017
     sadness       1.00      1.00      1.00    193437
    surprise       1.00      0.99      1.00     48729
       trust       1.00      0.99      0.99    205478

    accuracy                           1.00   1455563
   macro avg       1.00      1.00      1.00   1455563
weighted avg       1.00      1.00      1.00   1455563



In [19]:
test_pred = {'id':test_data_df.tweet_id, 'emotion':Y_test_pred}
test_pred = pd.DataFrame(test_pred)
test_pred.to_csv('Prediction.csv', index=False)

#### Multi-layer Perceptron

In [17]:
from sklearn.neural_network import MLPClassifier

X_train = train_data_BOW_features
Y_train = train_data_df.emotion

X_test = test_data_BOW_features

clf_MLP = MLPClassifier(hidden_layer_sizes=(200,100,50), activation='relu', solver='adam', batch_size='auto',
                        learning_rate_init=0.01, max_iter = 2000, shuffle=True, tol=1e-3, random_state=1, early_stopping=True)
clf_MLP = clf_MLP.fit(X_train, Y_train)

Y_train_pred = clf_MLP.predict(X_train)
Y_test_pred = clf_MLP.predict(X_test)

C:\Users\popo6\Anaconda3\envs\datamining\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:699: UserWarning: Training interrupted by user.
  warnings.warn("Training interrupted by user.")


NameError: name 'clf_decisiontree' is not defined

In [18]:
Y_train_pred = clf_MLP.predict(X_train)
Y_test_pred = clf_MLP.predict(X_test)

In [19]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

print(accuracy_score(y_true=Y_train, y_pred=Y_train_pred))
print(classification_report(y_true=Y_train, y_pred=Y_train_pred))

0.63872329813275
              precision    recall  f1-score   support

       anger       0.71      0.27      0.39     39867
anticipation       0.71      0.66      0.68    248935
     disgust       0.51      0.57      0.54    139101
        fear       0.66      0.56      0.61     63999
         joy       0.63      0.86      0.73    516017
     sadness       0.57      0.58      0.57    193437
    surprise       0.87      0.27      0.41     48729
       trust       0.82      0.34      0.48    205478

    accuracy                           0.64   1455563
   macro avg       0.69      0.51      0.55   1455563
weighted avg       0.66      0.64      0.62   1455563



In [20]:
test_pred = {'id':test_data_df.tweet_id, 'emotion':Y_test_pred}
test_pred = pd.DataFrame(test_pred)
test_pred.to_csv('Prediction_MLP.csv', index=False)

#### Deep Learning use keras

In [17]:
X_train = train_data_BOW_features
Y_train = train_data_df.emotion

X_test = test_data_BOW_features

In [18]:
import tensorflow.keras as keras
from sklearn.preprocessing import LabelEncoder
import numpy as np

label_encoder = LabelEncoder()
label_encoder.fit(Y_train)
print('check label: ', label_encoder.classes_)
print('\n## Before convert')
print('y_train[0:4]:\n', Y_train[0:4])
print('\ny_train.shape: ', Y_train.shape)

def label_encode(le, labels):
    enc = le.transform(labels)
    return keras.utils.to_categorical(enc)

def label_decode(le, one_hot_label):
    dec = np.argmax(one_hot_label, axis=1)
    return le.inverse_transform(dec)

y_train = label_encode(label_encoder, Y_train)

print('\n\n## After convert')
print('y_train[0:4]:\n', y_train[0:4])
print('\ny_train.shape: ', y_train.shape)

check label:  ['anger' 'anticipation' 'disgust' 'fear' 'joy' 'sadness' 'surprise'
 'trust']

## Before convert
y_train[0:4]:
 0             joy
1    anticipation
2             joy
3             joy
Name: emotion, dtype: object

y_train.shape:  (1455563,)


## After convert
y_train[0:4]:
 [[0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0.]]

y_train.shape:  (1455563, 8)


In [19]:
# I/O check
input_shape = X_train.shape[1]
print('input_shape: ', input_shape)

output_shape = len(label_encoder.classes_)
print('output_shape: ', output_shape)

input_shape:  994692
output_shape:  8


In [20]:
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.layers import ReLU, Softmax

# input layer
model_input = Input(shape=(input_shape, ))  # input size
X = model_input

# 1st hidden layer
X_W1 = Dense(units=100)(X)  # 100
H1 = ReLU()(X_W1)

# 2nd hidden layer
H1_W2 = Dense(units=64)(H1)  # 64
H2 = ReLU()(H1_W2)

# output layer
H2_W3 = Dense(units=output_shape)(H2)  # 8 output size
H4 = Softmax()(H2_W3)

model_output = H4

# create model
model = Model(inputs=[model_input], outputs=[model_output])

# loss function & optimizer
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# show model construction
model.summary()

Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 994692)]          0         
_________________________________________________________________
dense (Dense)                (None, 100)               99469300  
_________________________________________________________________
re_lu (ReLU)                 (None, 100)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)                6464      
_________________________________________________________________
re_lu_1 (ReLU)               (None, 64)                0         
_________________________________________________________________
dense_2 (Dense)              (None, 8)                 520       
_________________________________________________________________
softmax (Softmax)            (None, 8)                 0     

In [35]:
from tensorflow.keras.callbacks import CSVLogger

csv_logger = CSVLogger('training_log.csv')

# training setting
epochs = 6
batch_size = 32

# training!
history = model.fit(X_train, y_train, 
                    epochs=epochs, 
                    batch_size=batch_size, 
                    callbacks=[csv_logger])
print('training finish')

Epoch 1/6
45487/45487 [==============================] - 1581s 35ms/step - loss: 0.3302 - accuracy: 0.8906
Epoch 2/6
45487/45487 [==============================] - 1526s 33ms/step - loss: 0.3065 - accuracy: 0.8984
Epoch 3/6
45487/45487 [==============================] - 1521s 33ms/step - loss: 0.2885 - accuracy: 0.9045 - loss: 0.2885 - ac
Epoch 4/6
45487/45487 [==============================] - 1522s 33ms/step - loss: 0.2739 - accuracy: 0.9093
Epoch 5/6
45487/45487 [==============================] - 1521s 33ms/step - loss: 0.2609 - accuracy: 0.9137
Epoch 6/6
45487/45487 [==============================] - 1523s 33ms/step - loss: 0.2497 - accuracy: 0.91741:01 - loss: 0.2486 - accu
training finish


In [36]:
## predict
Y_test_pred = model.predict(X_test, batch_size=128)
Y_test_pred[:5]

array([[5.30865769e-08, 9.96636391e-01, 1.86843550e-07, 2.56285615e-09,
        1.03184127e-03, 1.97518489e-06, 9.79381002e-06, 2.31968984e-03],
       [1.06304565e-08, 9.15955976e-02, 2.52464565e-06, 1.63969034e-05,
        8.13233036e-09, 5.65887922e-06, 1.34613627e-04, 9.08245265e-01],
       [3.79486079e-03, 1.32101148e-01, 2.08835816e-03, 1.35375308e-02,
        2.60755420e-01, 5.86686313e-01, 7.43859098e-04, 2.92486919e-04],
       [8.56897736e-04, 4.90468264e-01, 9.51050388e-05, 3.09741794e-04,
        3.00378054e-01, 2.42723436e-05, 3.50386440e-03, 2.04363823e-01],
       [2.26804535e-04, 5.35527587e-01, 7.17054296e-04, 2.70781443e-02,
        2.95014004e-03, 2.79688742e-02, 2.00747736e-02, 3.85456592e-01]],
      dtype=float32)

In [37]:
Y_test_pred = label_decode(label_encoder, Y_test_pred)
Y_test_pred[:5]

array(['anticipation', 'trust', 'sadness', 'anticipation', 'anticipation'],
      dtype=object)

In [38]:
Y_train_pred = model.predict(X_train, batch_size=128)
Y_train_pred = label_decode(label_encoder, Y_train_pred)

In [39]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

print("Accuracy : "+str(accuracy_score(y_true=Y_train, y_pred=Y_train_pred)))
print(classification_report(y_true=Y_train, y_pred=Y_train_pred))

Accuracy : 0.932705077004568
              precision    recall  f1-score   support

       anger       0.96      0.83      0.89     39867
anticipation       0.94      0.94      0.94    248935
     disgust       0.93      0.93      0.93    139101
        fear       0.95      0.89      0.92     63999
         joy       0.92      0.97      0.94    516017
     sadness       0.92      0.93      0.93    193437
    surprise       0.95      0.86      0.90     48729
       trust       0.94      0.90      0.92    205478

    accuracy                           0.93   1455563
   macro avg       0.94      0.91      0.92   1455563
weighted avg       0.93      0.93      0.93   1455563



In [40]:
test_pred = {'id':test_data_df.tweet_id, 'emotion':Y_test_pred}
test_pred = pd.DataFrame(test_pred)
test_pred.to_csv('Prediction_keras.csv', index=False)

### 3.Third : A report of my work developping in the model for the competition

* I divide three part to do, data classification, data process and different way to train model.
    - In data classification part, use pandas to read the three files including “tweets_DM.json”, “emotion.csv” and “data_identification.csv”.
        1. In file “tweets_DM.json” we have interest in _source column and in that column we need to separate “tweet_id” ,”hashtags” and “text”. Additionally, I preserve information of score.
        2. In file “data_indentification.csv” I separate the put the same information to a pandas.dataframe according to train ant test. Then I use the method(isin) of pandas.dataframe to distinguish training data and testing data and put together respectively. So, now we have training data and testing data separately.
        3. Because we need to discriminate what sentiment it is according a tweet information, we have file “emotion.csv” that is for training data. Use previous processed training data and emotion.csv, then sort them in terms of “tweet_id” and concatenate two dataframe.
        4. Finally we have preliminary data about train and test. 
    
    - In data process part, I try to let the data is balance and tokenize text to train model.
        1. I think balanced data to train model is better than unbalanced data and the training data up to 1.4 million records. So, I don’ use oversampling but use undersampling to get data balanced.
        2. Then, use package countvertorizer to token text let me to get word information. I found it have one hundred thousand features. It’s so big. I will try to reduce them to train and compare accuracy of different features.

    - In train model, I use three kinds of model, decision tree, multi-layer perceptron using sklearn and neural network using keras, to compare and predict. Then use some package to see the performance matrix

        1. I use decisiontree model by package sklearn to train model, I change the parameter “max_depth” to compare the result. I set 100, 1000, 10000, 30000 and unlimit it. In testing data, the best result is 0.35732 by max_depth equal to 100. The other results are lower. Otherwise, I don’t reduce features and use all training data in previous results. Now, I reduce features from one hundred thousand to fifty thousand but I get worse results.

        2.  Then I use neural network by package keras to train model and predict. The parameters set same as TA teached, the testing score I get 0.4294. It improves obviously. I try to set different parameter, like the number of hidden layers, size of hidden layers, epochs and batch size. I can get the best score 0.4429 when two hidden layer about 100 and 64 neurons, epochs equal to 8 and batch size equal to 32. The same is that I use all training data and not reduce features to train. Now, I try to balance data using undersampling according to the smallest number of anger class but I get the worst score 0.30327.

        3.  Finally, I also use multi-layer perceptron by package sklearn. It take me a lot of time to train, I interrupt the training model when it train 5471 minutes. I don’t sure if I set very strict parameters let it unable to complete the training. Surprisingly, it gets the highest score 0.46807 although I use all training data and all features. It takes too much time, I can’t train it using different parameters. 
* Conclusion :
    1.  If I use all training data and all feature to train, it can have a good result. Moreover, multi-layer perceptron by sklearn is best model in three kinds of model but it takes too much time. Maybe the parameters I set in others model are not well, I think there must be more suitable parameters to get the greater results.

    2.  On the other hand, in data process part I try to undersampling data to get balanced data. But the number of anger is too lower, the remaining training data is too low even lower than testing data. I should process balanced data using better method, maybe as balanced as possible without throwing away data too much. Or setting a threshold , then number exceeds threshold to reduce number, on the contrary, duplicate data or do similar fake data.

    3.  In features dimension, it is too higher. I should do some method to handle the text data. Close to deadline, I found that there are some unncecessary information in the text like URL. If I can delete those, maybe dimension will reduce greatly. In addition, using different word processing might allow the model to train better. Maybe try word2vec to make words related to each other, I guess it will make model train better.
